### PR TITLE
Add support for bias conversion in Swiglu models

### PR DESCRIPTION
**Pull request by titu1994**
# What does this PR do ?

Adds support for bias conversion in TP PP Script for swiglu supported modls

**Collection**: [NLP]

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [x] Bugfix
- [ ] Documentation

### CODE REVIEW

**Review by odelalleau — COMMENTED** [review]
Looks good to me, but do you know why `linear_fc1` only appears in the code that does the merge and not in the code that does the split?

(I realize it was the case before this PR, but it seems a bit weird at first glance)

File: examples/nlp/language_modeling/megatron_change_num_partitions.py [hunk 1]
```diff
@@ -1384,6 +1385,38 @@ def main():
                 with open_dict(model.cfg):
                     model.cfg.tokenizer.model = args.tokenizer_model_path
 
+            else:
+                # Extract all artifacts that are not binary files
+                # and save them to the model directory
```
Comment (odelalleau) [inline examples/nlp/language_modeling/megatron_change_num_partitions.py:1390]: What's the motivation for this change? This seems unrelated to #7383

File: examples/nlp/language_modeling/megatron_change_num_partitions.py [hunk 2]
```diff
@@ -1384,6 +1385,38 @@ def main():
...
+                                continue
+
+                            extracted_file = tar.extractfile(member)
+                            extracted_file_path = os.path.join(temp_dir, member.name)
+
+                            if '.model' in member.name:
+                                tokenizer_model_path = extracted_file_path
+
+                            with open(extracted_file_path, "wb") as f:
+                                f.write(extracted_file.read())
```
Comment (odelalleau) [inline examples/nlp/language_modeling/megatron_change_num_partitions.py:1410]: Besides the tokenizer ".model" file, are there other files required by the tokenizer that need to be extracted? (at first glance it seems like we're not using any of the other extracted files)

File: examples/nlp/language_modeling/megatron_change_num_partitions.py [hunk 3]
```diff
@@ -1384,6 +1385,38 @@ def main():
                 with open_dict(model.cfg):
                     model.cfg.tokenizer.model = args.tokenizer_model_path
 
+            else:
+                # Extract all artifacts that are not binary files
+                # and save them to the model directory
+                logging.info("Extracting artifact from NeMo file...")
```
Comment (odelalleau) [inline examples/nlp/language_modeling/megatron_change_num_partitions.py:1391]: Could this logic be moved outside of the `for pp_rank` loop so that it's run only once?

(also, ideally what comes after the extraction code would be in a `try...finally` block so that `temp_dir` is always deleted on exit)

Comment (titu1994) [reply examples/nlp/language_modeling/megatron_change_num_partitions.py:1391]: Not really. This logic is shared for all TP PP and must execute in a loop.
It should be fast. The extraction only works on the base level of the nemo file and does not recurse the actual checkpoint dirs.

File: examples/nlp/language_modeling/megatron_change_num_partitions.py (hunk 2 above)
Comment (odelalleau) [reply examples/nlp/language_modeling/megatron_change_num_partitions.py:1410]: In that case, if only the ".model" file is needed, then why not only extract this file?

Comment (titu1994) [reply examples/nlp/language_modeling/megatron_change_num_partitions.py:1410]: There are models which dont have it, or have multiple. This is a fallback to not require the tokenizer all the time.

Comment (odelalleau) [reply examples/nlp/language_modeling/megatron_change_num_partitions.py:1410]: Sorry, I'm not sure I understand:
* If there's no ".model" file, you may end up extracting a bunch of files that will never be used
* If there are multiple ".model" files, you end up using only the last one (which I guess is an arbitrary choice) and extracting a bunch of other files that will never be used

Why not just extract at most one ".model" file and stop there? (skipping over anything that's not ".model")

Comment (titu1994) [reply examples/nlp/language_modeling/megatron_change_num_partitions.py:1410]: This only extracts at the root directory, not at checkpoint level.  So the cost of extraction is trivial and order of kb or mb for huge Tokenizer. 

For either of those cases, if it has no Tokenizer it will fail when calling the constructor of the model - it always needs one tokenizer. For multiple tokenizer - this code path isn't supported at the moment. It's arbitrary yes, and will still fail when loading the constructor because such models require two Tokenizers and youre providing just one.

File: examples/nlp/language_modeling/megatron_change_num_partitions.py (hunk 3 above)
Comment (titu1994) [reply examples/nlp/language_modeling/megatron_change_num_partitions.py:1391]: No. Simply calling tarfile open does not untar the file - it only opens a object that can traverse the tar object, not it's contents inside it. You need to explicitly call member.extractfile() to get the file pointer to the actual object

File: examples/nlp/language_modeling/megatron_change_num_partitions.py (hunk 2 above)
Comment (odelalleau) [reply examples/nlp/language_modeling/megatron_change_num_partitions.py:1410]: Thanks for the clarification, but even if it's cheap, why extract multiple files when it's trivial to extract only the one you need? Note that (with my admittedly limited knowledge of NeMo) I see no reason why future models may not include very large files other than .ckpt or .pt...

File: examples/nlp/language_modeling/megatron_change_num_partitions.py (hunk 3 above)
Comment (odelalleau) [reply examples/nlp/language_modeling/megatron_change_num_partitions.py:1391]: I quoted that line to show that it was the same file, the actual extraction occurs below in the code. My point is that the same file(s) are extracted for each PP rank which seems pointless, it would be cleaner to do this only once before the `for` loop.

File: examples/nlp/language_modeling/megatron_change_num_partitions.py [hunk 4]
```diff
@@ -978,6 +978,30 @@ def main():
     if vp_size > 1:
         set_virtual_parallel_rank_safely(0)
 
+    # Extract tokenizer artifact from the model to temp directory
+    logging.info("Extracting tokenizer artifact from NeMo file...")
```
Comment (odelalleau) [inline examples/nlp/language_modeling/megatron_change_num_partitions.py:982]: Should probably be inside an
```python
if args.tokenizer_model_path is None:
```

**Review by odelalleau — COMMENTED** [review]
Thanks for the refactor! I think it's easier to follow the logic now :)

NB: you could delete `temp_dir` afterwards (not sure it's worth wrapping everything in a `try... finally` since it would cause an extra indentation, so just deleting it at the end would probably be enough)

**Review by odelalleau — APPROVED** [review]
Looks good to merge!

**Review by ericharper — APPROVED** [review]
LGTM. Thanks!
